rotkehlchen/chain/evm/decoding/rainbow/decoder.py

```diff
@@ -0,0 +1,388 @@
+
+    ) -> DecodingOutput:
+        """
+        Decodes a swap transaction from token to native token.
+        Check if the first log is transfer and the last log is withdrawn
+        """
+        native_token = self.evm_inquirer.native_token
+        unwrap_log = all_logs[-1]
+        if (
+            (all_logs[0] != ERC20_OR_ERC721_TRANSFER and bytes_to_address(all_logs[0].topics[2]) == sender) or  # noqa: E501
+            unwrap_log.topics[0] != WETH_WITHDRAW_TOPIC
+        ):
+            return DEFAULT_DECODING_OUTPUT
+
+        if not (in_event and out_event):
+            return DEFAULT_DECODING_OUTPUT
+
+        try:
+            in_amount = asset_raw_value(in_event.amount, native_token)
+        except UnsupportedAsset:
+            in_amount = 0
+        # The fee will be the withdraw data subtract the in_amount
+        fee_raw = int.from_bytes(unwrap_log.data) - in_amount
+
+        tx_hash = transaction.tx_hash
+        # Need to change the index of receive to respect the events order
+        in_event.sequence_index = out_event.sequence_index + 1
+
+        # Create the event for the fee
+        fee_event = self._create_fee_event(
+            tx_hash=tx_hash,
+            sequence_index=out_event.sequence_index + 2,
+            timestamp=transaction.timestamp,
+            fee_raw=fee_raw,
+            fee_asset=native_token,
+            sender=sender,
+        )
+
+        return DecodingOutput(event=fee_event)
+
+    def _maybe_decode_swap_token_to_token(
+        self,
+        transaction: EvmTransaction,
+        decoded_events: list['EvmEvent'],
+        all_logs: list[EvmTxReceiptLog],
+        sender: ChecksumEvmAddress,
+        out_event: Optional['EvmEvent'],
+    ) -> DecodingOutput:
+        """Decodes a swap transaction from token to native token.
+        Check if the first and last logs is transfer
+        """
+        if out_event is None:
+            return DEFAULT_DECODING_OUTPUT
+
+        sell_amount = swap_amount = None
+        for log in all_logs:
+            if (
+                log.address == out_event.asset.resolve_to_evm_token().evm_address and
+                log.topics[0] == ERC20_OR_ERC721_TRANSFER
+            ):
+                # Find the transfer log from the user wallet to router
+                if bytes_to_address(log.topics[2]) == self.router_address:
+                    sell_amount = int.from_bytes(log.data)
+
+                # Find the transfer log from the router to the swap
+                if bytes_to_address(log.topics[1]) == self.router_address:
+                    swap_amount = int.from_bytes(log.data)
+
+                if swap_amount and sell_amount:
+                    break
+
+        if not (sell_amount and swap_amount):
+            return DEFAULT_DECODING_OUTPUT
+
+        # Remove the transfer event
+        # of the purchased token to the user's wallet
+        # to create a new event with the right sequence index
+        transfer_log = all_logs[-1]
```

Comment:
don't rely on this as it is error prone. Identify the events by checking that is a transfer and the to/from